File: SOLID/code-examples/chapter-05.py
# FILE: chapter-05.py
# Chapter 5 — Polymorphism — Python

from abc import ABC, abstractmethod
import math

# Base class with abstract methods
class Shape(ABC):
    @abstractmethod
    def get_area(self):
        pass
    
    @abstractmethod
    def draw(self):
        pass

# Derived classes
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius
    
    def get_area(self):
        return math.pi * self.radius ** 2
    
    def draw(self):
        print(f"Drawing a circle with radius {self.radius}")

class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def get_area(self):
        return self.width * self.height
    
    def draw(self):
        print(f"Drawing a rectangle {self.width}x{self.height}")

def main():
    print("🔄 Polymorphism Example (Python)")
    print("=================================\n")
    
    # Create different shapes
    shapes = [
        Circle(5.0),
        Rectangle(4.0, 6.0),
        Circle(3.0)
    ]
    
    # Polymorphism in action - same interface, different behavior
    for shape in shapes:
        shape.draw()                           # Calls appropriate draw method
        print(f"Area: {shape.get_area():.2f}\n")  # Calls appropriate get_area method
    
    print("💡 Polymorphism Benefits:")
    print("   ✓ Same interface for different objects")
    print("   ✓ Runtime method resolution")
    print("   ✓ Extensible - can add new shapes easily")

if __name__ == "__main__":
    main()

File: SOLID/code-examples/chapter-05.cpp
// FILE: chapter-05.cpp
// Chapter 5 — Polymorphism — C++

#include <iostream>
#include <vector>
#include <memory>
using namespace std;

// Base class with virtual methods
class Shape {
public:
    virtual double getArea() = 0;  // Pure virtual method
    virtual void draw() = 0;       // Pure virtual method
    virtual ~Shape() = default;    // Virtual destructor
};

// Derived classes
class Circle : public Shape {
private:
    double radius;
    
public:
    Circle(double r) : radius(r) {}
    
    double getArea() override {
        return 3.14159 * radius * radius;
    }
    
    void draw() override {
        cout << "Drawing a circle with radius " << radius << endl;
    }
};

class Rectangle : public Shape {
private:
    double width, height;
    
public:
    Rectangle(double w, double h) : width(w), height(h) {}
    
    double getArea() override {
        return width * height;
    }
    
    void draw() override {
        cout << "Drawing a rectangle " << width << "x" << height << endl;
    }
};

int main() {
    cout << "🔄 Polymorphism Example (C++)" << endl;
    cout << "=============================" << endl << endl;
    
    // Create different shapes
    vector<unique_ptr<Shape>> shapes;
    shapes.push_back(make_unique<Circle>(5.0));
    shapes.push_back(make_unique<Rectangle>(4.0, 6.0));
    shapes.push_back(make_unique<Circle>(3.0));
    
    // Polymorphism in action - same interface, different behavior
    for (const auto& shape : shapes) {
        shape->draw();                    // Calls appropriate draw method
        cout << "Area: " << shape->getArea() << endl << endl;  // Calls appropriate getArea method
    }
    
    cout << "💡 Polymorphism Benefits:" << endl;
    cout << "   ✓ Same interface for different objects" << endl;
    cout << "   ✓ Runtime method resolution" << endl;
    cout << "   ✓ Extensible - can add new shapes easily" << endl;
    
    return 0;
}